wallet: Remove return value from CommitTransaction

jnewbery opened the pull request:
`CommitTransaction()` returns a bool to indicate success, but since commit
b3a7410 (#9302) it only returns true, even if the transaction was not
successfully broadcast. This commit changes CommitTransaction() to return
void.

All dead code in `if (!CommitTransaction())` branches has been removed.

Two additional commits fix up the idiosyncratic whitespace in `CommitTransaction` and add a doxygen comment for the function.

jnewbery committed:
[wallet] Fix whitespace in CWallet::CommitTransaction()

Reviewer hint: use --ignore-all-space git diff option for review.

jnewbery committed:
[wallet] Add doxygen comment to CWallet::CommitTransaction()

jnewbery committed:
[wallet] Remove return value from CommitTransaction()

CommitTransaction returns a bool to indicate success, but since commit
b3a74100b8 it only returns true, even if the transaction was not
successfully broadcast. This commit changes CommitTransaction() to return
void.

All dead code in `if (!CommitTransaction())` branches has been removed.

MarcoFalke commented:
Slightly tend to NACK. What is this needed for?

Effectively, a gui user will send the transaction twice (spending different coins), losing all their funds.

jnewbery commented:
@MarcoFalke 
> Slightly tend to NACK. What is this needed for?

Currently the function returns a bool, which is always `true`, but the callers have logic that depends on the return value.

`CommitTransaction()` did previously return `true`/`false` based on success. That was removed in #9302 so that the `send___` RPC methods would return the txid even if the transaction failed to broadcast. That's useful because the transaction will still exist in the wallet, so the RPC user would want to know about it. The fix should have been to have the call sites ignore the return value of `CommitTransaction()`, but instead the PR changed the function itself to not return success/failure, meaning that as a side effect `WalletModel::sendCoins()` is no longer informed if transaction broadcast fails. From reading the PR, it looks to me like that change was unintentional.

I could change this PR to maintain the current behaviour (not return success to `WalletModel::sendCoins()`). Do you think that's preferable?

MarcoFalke commented:
> From reading the PR, it looks to me like that change was unintentional.

I am pretty sure it was intentional. At least that was my understanding, see https://github.com/bitcoin/bitcoin/pull/9302#issuecomment-266284389

To sum up, the wallet is only allowed to do one of the following:
* Create a tx, sign it and add it to the wallet, so that it can be rebroadcast immediately or later, then `return true`
* Create a tx, sign it and throw it away for some reason, then `return false`

A wallet is not allowed to create a tx, sign it and add it to the wallet, so that it will be rebroadcast later, but `return false`. <-- This is what you pull request is doing.

jnewbery commented:
In that case, `CommitTransaction` should not return anything, and all the dead code in `if (!CommitTransaction())` blocks should be removed.

MarcoFalke commented:
Fine with me

jnewbery commented:
This PR now just changes `CommitTransaction` to return `void`, and deletes all dead code.

Also rebased on master.

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -1147,7 +1147,18 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
+     * @param reserveKey[in,out] keypool key used for change
+     * @param state[in,out] CValidationState object returning information about whether the transaction was accepted
+     */
+void CommitTransaction(CTransactionRef tx, mapValue_t mapValue, std::vector<std::pair<std::string, std::string>> orderForm, CValidationState& state);
```
```suggestion
    void CommitTransaction(CTransactionRef tx, mapValue_t mapValue, std::vector<std::pair<std::string, std::string>> orderForm, CValidationState& state);
```

:eyes:

jnewbery replied:
fixed

achow101 commented on src/wallet/wallet.h:
```diff
@@ -1147,6 +1147,17 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
+     * @param tx[in,out] The transaction to be broadcast.
+     * @param mapValue[in] key-values to be set on the transaction.
+     * @param orderForm[in] BIP 70 / BIP 21 order form details to be set on the transaction.
+     * @param reserveKey[in,out] keypool key used for change
```
`reserveKey` is not an argument for this function.

jnewbery replied:
fixed

achow101 reviewed (commented):
Code Review ACK 5021b8a7fb4de518155146de6cac231e91e874f8

jnewbery commented:
Addressed @achow101 's review comment (thanks!)

achow101 commented:
ACK 1f490c9a5f8767477f6ffe382b4a0578dbac5e3e

promag commented:
Concept ACK.

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -1147,6 +1147,16 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
+     * Should be called after CreateTransaction unless you want to abort
+     * broadcasting the transaction.
+     *
+     * @param tx[in,out] The transaction to be broadcast.
```
in commit c7397e074996086578d9ac3db78d654fdbbf58a5:

this is not a return parameter. The tx is consumed (`std::move`)

jnewbery replied:
Thanks. Fixed.

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -1147,6 +1147,16 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
+     * @param tx[in,out] The transaction to be broadcast.
+     * @param mapValue[in] key-values to be set on the transaction.
+     * @param orderForm[in] BIP 70 / BIP 21 order form details to be set on the transaction.
+     * @param state[in,out] CValidationState object returning information about whether the transaction was accepted
```
I think this should just be `[out]`?

jnewbery replied:
This actually isn't set any more (since commit 611291c198). I've removed it in a new commit.

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -1157,7 +1157,7 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
      * @param state[in,out] CValidationState object returning information about whether the transaction was accepted
      */
-    bool CommitTransaction(CTransactionRef tx, mapValue_t mapValue, std::vector<std::pair<std::string, std::string>> orderForm, CValidationState& state);
+    void CommitTransaction(CTransactionRef tx, mapValue_t mapValue, std::vector<std::pair<std::string, std::string>> orderForm, CValidationState& state);
```
in commit 1f490c9a5f8767477f6ffe382b4a0578dbac5e3e you claim that the return value is removed, but you still return `state`, which is used in a single place (`feebumper`). Haven't thought too much about this, but it seems fine to use there because it can't lead to loss of funds?

Also, that code path shouldn't be hit unless there is a sudden massive fee spike or a logic error.

jnewbery replied:
I've removed `state`, since it's never actually set.

MarcoFalke replied:
ouch. Good catch :grimacing:

jnewbery committed:
[wallet] Remove `state` argument from CWallet::CommitTransaction

The `state` return argument has not been set since commit 611291c198.
Remove it (and the one place that it's used in a calling function).

MarcoFalke commented:
Concept ACK

laanwj commented:
ACK 9e95931865186d7a9a6dc54b64bd96507e9fea4b

JeremyRubin commented:
Hmmm -- sorry for a posthumous comment here, but this patch confuses me a little.

It seems the better change would be to continue to have CommitTransaction return a success value/reject reason.

What's the recommended pattern following this change for checking the status of a committed transaction? What if a transaction is rejected for consensus reasons, and won't be in the wallet? As a wallet rpc implementer, I want to be able to detect a failed committransaction and then undo some prior steps...

jnewbery commented:
This PR didn't change any behaviour. It removed a return value that was not used by any calling code.

See #9302 for why the RPC code was changed.

MarcoFalke commented:
And the TODO that this should, is still in the code:

```cpp
     // TODO: if we expect the failure to be long term or permanent, instead delete wtx from the wallet and return failure.

JeremyRubin commented:
Yeah I read those; my point is that it seems that the patched behavior of the function is buggy since #9302, and then in #17154 we just enforce the buggy behavior rather than addressing the TODO and returning an error code, which users of the function may prefer to have...

jnewbery commented:
Contributions are welcome. If you want to suggest a PR to return failure information from `CommitTransaction()`, please go ahead.

MarcoFalke commented:
Instead of creating a transaction that is consensus-invalid and then returning a failure and removing the tx again, I'd prefer if the consensus-invalid transaction was not created by the wallet in the first place. Or if that is not possible, the wallet should have checked consensus-validity before storing it in the wallet.dat

JeremyRubin commented:
@MarcoFalke well I don't disagree it's better to never be wrong in the first place! But the committransaction also seems to be able to fail for mere policy reasons too, right?


@jnewbery My original question was, in light of this patch, if there is actually a recommended handling for transactions which may err, what's the "proper handling" to recover the relevant information here?

I'd be happy to make a PR which makes the changes but I don't really fully understand what we want functionality wise from CommitTransaction.

MarcoFalke commented:
> also seems to be able to fail for mere policy reasons too, right

Yes.
If the policy fail is expected to be short term, the wallet should continue as usual and commit the tx to the `wallet.dat`. If the policy fail is expected to be long term or permanent, the wallet shouldn't commit the tx in the first place.

I think you are arguing that the functionality to check the tx should happen in `CommitTransaction` and that this pull request made it harder to do so. I think we can only know best where to put this check after someone has implemented it. It might be better as a standalone function `PreCheckCommitTransaction` or similar, idk.

> a recommended handling for transactions which may err

I think we don't have a recommendation to recover from such errors. The best we can do right now is manual fixups: `abandontransaction` (maybe in combination with disabled wallet rebroadcast).

jnewbery commented:
I think the most critical thing in `CommitTransaction()` is that the transaction is added to the wallet and old coins are marked spent _before_ the transaction is broadcast. It would be a problem if the transaction was broadcast and then failed to be added to the wallet.

I think adding a call to ATMP with `test_accept` before calling `AddToWallet()` and then returning any error could be an improvement.
